### ISSUE TITLE
DataChannel Blob() suggestion

### ISSUE THREAD

**Issue by subversivo58** [node-webrtc/node-webrtc#363]
I was looking for [lib/datachannel.js line 81](https://github.com/js-platform/node-webrtc/blob/develop/lib/datachannel.js#L81) when I came across the comment:

> Note: not sure what to do about this, since node doesn't have a Blob API

Would not it be a matter of converting a `Blob()` to `base64` in the front-end and sending an `{Object}` ... and in the back-end `Buffer()`?

**front-end**

```javascript
// convert
let blobToBase64 = function(blob, cb) {
    let reader = new FileReader()
    reader.onload = function() {
        let dataUrl = reader.result,
             base64 = dataUrl.split(',')[1]
        cb(base64)
    }
    reader.readAsDataURL(blob)
}
// use
blobToBase64(blob, function(base64) {
    // send an {Object} with the necessary information
   let blobToSend = {
        data: base64,
        type: 'mime/type'
   }
   // send ...
}
```

In the back-end it would simply create a `Buffer()` of this raw data encapsulated in `base64`, could use additional keys like **mime-type** and **size** within the resulting stream

**back-end**

```javascript
// assuming that the blob variable is the {Object} coming from the front-end
let buf = new Buffer(blob.data, 'base64')
// ...
```

I am not sure as to the specification, maybe this approach is not correct but it seems to me feasible

Congratulations on the project, I wish you much success

**Comment by markandrus** [comment]
@subversivo58 we could probably do something like this, but we might want to know what kind of Blob (and possibly FileReader) implementation we have. AFAIK, there's really only one "working" Blob implementation for Node, and that's the one provided by [jsdom](https://github.com/jsdom/jsdom). (AFAICT, it's not terribly useful in Node... but it exists). Do you know of any others?

**Sending Blobs**

The FileReader API complicates sending Blobs, so it might make sense to "cheat" and read jsdom's Blob implementation's internal `_buffer` property. I actually landed some code to do this in https://github.com/js-platform/node-webrtc/commit/4a6d751b12a56d21cca03230fa8c926762bb7094.

**Receiving Blobs**

Constructing one of these things from an ArrayBuffer is easy enough if we have the Blob constructor, but the question is, "How do we know what Blob constructor to use?".

* Should we just look up a global if it's available?
* Should we try to `require('jsdom')` internally?
* If we did that, would it create problems with mixed versions of jsdom? 
* Should we instead pass the Blob constructor via ExtendedRTCConfiguration? e.g., `new RTCPeerConnection({ iceServers: [], Blob })`?

Any feedback you or the rest of the community has here would be super helpful. 👍

**Comment by subversivo58** [comment]
I was looking at the documentation and `new Buffer()` was deprecated instead of the correct one is to use `Buffer.from()`

https://nodejs.org/api/buffer.html#buffer_class_method_buffer_from_buffer

When I suggested this it would be something like: use in a high-level API on the front end for the back end (or WebRTC backbone implementation) ... using some flag the backend might know that it is receiving this "schema "of the front-end (just because it does not have a native function) a sign in the opposite direction should also occur. The front end might recognize that it is receiving a backbone buffer.

I'm not sure how this could (or should) look like, I have no test case to substantiate my argument.


From browser to browser the normal use of `Blob()` , from browser to backbone uses `FileReader()` and between backbone nodes `Buffer()` plus both ends should be able to know who they are dealing with in order to handle the received format.

The JSDom library seems to bring the closest approach, but looking from the outside just as an observer: I do not think it right to enclose a dependency.

I found the last item on your list elegant, but I believe a flag should come with a proposed approach:

* how the backbone identifies the resource type (whether it came from the front-end or from another backbone)
* how the browser identifies the resource type (if it came from another browser or from a backbone)

Knowing this, I believe it is half way to a satisfactory approach to implement the feature in a cross-browser way.

Basically all browsers that support `RTCPeerConnection` also support `FileReader`. No?

**Comment by piranna** [comment]
`ws` module [don't support `blob` for the binaryType](https://github.com/websockets/ws/blob/master/doc/ws.md#websocketbinarytype) field, but instead it provides a custom `nodebuffer` type that generates a Node.js `Buffer` object. It's not standard, but it's explicit, so I don't find it so bad. Obviously it also can generate `arrayBuffer` objects, that seems to be the only one that can be used to create both Node.js and browser compatible code...
